modules/core/lib/Auth/Process/SubjectID.php

```diff
@@ -0,0 +1,167 @@
+
+    /**
+     * The regular expression to match the specifications
+     *
+     * @var string
+     */
+    public const SPEC_PATTERN = '/^[a-z0-9][a-z0-9=-]{0,126}@[a-z0-9][a-z0-9.-]{0,126}$/i';
+
+    /**
+     * The regular expression to match worrisome identifiers that need to raise a warning
+     *
+     * @var string
+     */
+    public const WARN_PATTERN = '/^[a-z0-9][a-z0-9=-]{3,126}@[a-z0-9][a-z0-9.-]{3,126}$/i';
+
+    /**
+     * The attribute we should generate the subject id from.
+     *
+     * @var string
+     */
+    protected string $identifyingAttribute;
+
+    /**
+     * The scope to use for this attribute.
+     *
+     * @var string
+     */
+    protected string $scope;
+
+
+    /**
+     * Initialize this filter.
+     *
+     * @param array &$config  Configuration information about this filter.
+     * @param mixed $reserved  For future use.
+     */
+    public function __construct(array &$config, $reserved)
+    {
+        parent::__construct($config, $reserved);
+
+        Assert::keyExists($config, 'identifyingAttribute', "Missing mandatory 'identifyingAttribute' config setting.");
+        Assert::keyExists($config, 'scope', "Missing mandatory 'scope' config setting.");
+        Assert::stringNotEmpty($config['identifyingAttribute']);
+        Assert::regex(
+            $config['scope'],
+            self::SCOPE_PATTERN,
+            'core:' . static::NAME . ': \'scope\' contains illegal characters.'
+//            ProtocolViolationException::class
+        );
+
+        $this->identifyingAttribute = $config['identifyingAttribute'];
+        $this->scope = $config['scope'];
+    }
+
+
+    /**
+     * Apply filter to add the subject ID.
+     *
+     * @param array &$state  The current state.
+     */
+    public function process(array &$state): void
+    {
+        $userID = $this->getIdentifyingAttribute($state);
+
+        $value = strtolower($userID . '@' . $this->scope);
+        $this->validateGeneratedIdentifier($value);
+
+        $state['Attributes'][Constants::ATTR_SUBJECT_ID] = [$value];
+    }
+
+
+    /**
+     * Retrieve the identifying attribute from the state and test it for erroneous conditions
+     *
+     * @param array $state
+     * @return string
+     * @throws \SimpleSAML\Assert\AssertionFailedException if the pre-conditions are not met
+     */
+    protected function getIdentifyingAttribute(array $state): string
+    {
+        Assert::keyExists($state, 'Attributes');
+        Assert::keyExists(
+            $state['Attributes'],
+            $this->identifyingAttribute,
+            sprintf(
+                "core:" . static::NAME . ": Missing attribute '%s', which is needed to generate the ID.",
+                $this->identifyingAttribute
+            )
+        );
+
+        $userID = $state['Attributes'][$this->identifyingAttribute][0];
+        Assert::notEmpty($userID, 'core' . static::NAME . ': \'identifyingAttribute\' cannot be an empty string.');
```

Comment:
is it even better to use `Assert::stringNotEmpty`?

Reply:
Agreed on both the docs and the assertion!

Reply:
Any remarks on the docs?